altsrc/detect_input_source.go

```diff
@@ -0,0 +1,43 @@
+package altsrc
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/urfave/cli/v2"
+)
+
+var detectableSources = map[string]func(string) func(cCtx *cli.Context) (InputSourceContext, error){
```

Comment:
Trying to envision how it might be used in parallel since it resolves to an actual `InputSourceContext` before the app (or command/subcommand) even runs.

What should I attach it to so it stays in the `altsrc` package?

Reply:
(Unless I can pull in a dependency on `github.com/orcaman/concurrent-map/v2`, at which point it doesn't have to belong to anything.)

Reply:
You could move it out of altsrc and attach it to app.

Reply:
Just gotta work out how to do it without introducing a circular dependency.

Reply:
Opted for a hybrid approach with a read-only map for the default/built-in types, and a read-write map attached to `App` for developer-defined types and overrides. The read-write map (`detectableSources`) is accessed through public methods on `App` itself. I left most of the logic in `altsrc` to avoid circular dependencies, and for consistency with the other types.